### PR TITLE
VMware: New module vmware_vm_host_drs_rule

### PR DESCRIPTION
**Pull request by karstenjakobsen**
##### SUMMARY
New module to create DRS VM-HOST rules in a given cluster. 

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
vmware_vm_host_drs_rule

##### ADDITIONAL INFORMATION
This module complements
```vmware_drs_group``` https://github.com/ansible/ansible/pull/50650
```vmware_drs_group_facts``` https://github.com/ansible/ansible/pull/50649

### CODE REVIEW

File: lib/ansible/modules/cloud/vmware/vmware_vm_host_drs_rule.py [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+    description:
+      - "If set to C(True), the DRS rule will be an Affinity rule."
+      - "If set to C(False), the DRS rule will be an Anti-Affinity rule."
+      - "Effective only if C(state) is set to C(present)."
+    type: bool
+  datacenter:
+    aliases:
+      - datacenter_name
+    description:
+      - "Datacenter to search for given cluster. If not set, we use first cluster we encounter with C(cluster_name)"
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_vm_host_drs_rule.py]:
```suggestion
      - "Datacenter to search for given cluster. If not set, we use first cluster we encounter with C(cluster_name)."
```

File: lib/ansible/modules/cloud/vmware/vmware_vm_host_drs_rule.py [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+    required: false
+  cluster_name:
+    description:
+      - "Cluster to create VM-Host rule"
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_vm_host_drs_rule.py]:
```suggestion
      - "Cluster to create VM-Host rule."
```

File: lib/ansible/modules/cloud/vmware/vmware_vm_host_drs_rule.py [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+    required: true
+  drs_rule_name:
+    description:
+      - "Name of rule to create or remove"
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_vm_host_drs_rule.py]:
```suggestion
      - "Name of rule to create or remove."
```

File: lib/ansible/modules/cloud/vmware/vmware_vm_host_drs_rule.py [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+    required: true
+  enabled:
+    default: false
+    description:
+      - "If set to C(True), the DRS rule will be enabled."
+      - "Effective only if C(state) is set to C(present)."
+    type: bool
+  host_group_name:
+    description:
+      - "Name of Host group to use with rule"
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_vm_host_drs_rule.py]:
```suggestion
      - "Name of Host group to use with rule."
```

File: lib/ansible/modules/cloud/vmware/vmware_vm_host_drs_rule.py [hunk 5]
```diff
@@ -0,0 +1,452 @@
...
+      - present
+      - absent
+    default: present
+    description:
+      - "If set to C(present) and the rule doesn't exists then the rule will be created."
+      - "If set to C(absent) and the rule exists then the rule will be deleted."
+    required: true
+  vm_group_name:
+    description:
+      - "Name of VM group to use with rule"
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_vm_host_drs_rule.py]:
```suggestion
      - "Name of VM group to use with rule."
```

File: lib/ansible/modules/cloud/vmware/vmware_vm_host_drs_rule.py [hunk 6]
```diff
@@ -0,0 +1,452 @@
...
+    enabled: True
+    affinity_rule: True
+'''
+
+RETURN = r'''
+
+'''
+
+try:
+    from pyVmomi import vim, vmodl
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_vm_host_drs_rule.py]: `vmodl` is unused.

File: lib/ansible/modules/cloud/vmware/vmware_vm_host_drs_rule.py [hunk 7]
```diff
@@ -0,0 +1,452 @@
...
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils._text import to_native
+from ansible.module_utils.vmware import (PyVmomi, vmware_argument_spec, wait_for_task, find_cluster_by_name,
+                                         find_datacenter_by_name)
+
+
+class VmwareVmHostRuleDrs(PyVmomi):
+    """
+    Doctring
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_vm_host_drs_rule.py]:
```suggestion
    Class to manage VM HOST DRS Rules
```

File: lib/ansible/modules/cloud/vmware/vmware_vm_host_drs_rule.py [hunk 8]
```diff
@@ -0,0 +1,452 @@
...
+
+        # Dont populate lists if we are deleting group
+        if state == 'present':
+            # Get list of vm groups only if state is present
+            self.__vm_group_obj = self.__get_group_by_name(group_name=self.__vm_group_name)
+            self.__host_group_obj = self.__get_group_by_name(group_name=self.__host_group_name, host_group=True)
+
+    def get_msg(self):
+        """
+        Docstring
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_vm_host_drs_rule.py]: Might want to replace this with help message.

File: lib/ansible/modules/cloud/vmware/vmware_vm_host_drs_rule.py [hunk 9]
```diff
@@ -0,0 +1,452 @@
...
+
+                task = self.__cluster_obj.ReconfigureEx(config_spec, modify=True)
+                wait_for_task(task)
+
+            self.__changed = True
+
+        rule_obj = self.__get_rule_key_by_name(rule_name=self.__rule_name)
+        self.__result = self.__normalize_vm_host_rule_spec(rule_obj)
+
+        self.__msg = "Created DRS rule %s successfully" % (self.__rule_name)
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_vm_host_drs_rule.py]: Change this message when we are editing an existing DRS rule. Right now it is showing "Created ..." even if we are updating the DRS Rule.

File: lib/ansible/modules/cloud/vmware/vmware_vm_host_drs_rule.py [hunk 10]
```diff
@@ -0,0 +1,452 @@
...
+            config_spec = vim.cluster.ConfigSpecEx(rulesSpec=[rule_spec])
+
+            if not self.module.check_mode:
+
+                task = self.__cluster_obj.ReconfigureEx(config_spec, modify=True)
+                wait_for_task(task)
+
+            self.__changed = True
+
+        self.__msg = "Deleted DRS rule `%s` successfully" % (self.__rule_name)
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_vm_host_drs_rule.py]: For already deleted rule, we can change message like 

```python
else:
    self.__msg = "DRS Rule `%s` does not exists or already deleted" % (self.__rule_name)

```

**Review by Akasurde — CHANGES_REQUESTED** [review]
Rest LGTM

**Comment by Akasurde** [comment]
@karstenjakobsen Thanks for submitting this PR.

File: lib/ansible/modules/cloud/vmware/vmware_vm_host_drs_rule.py (hunk 9 above)
Comment (karstenjakobsen) [reply lib/ansible/modules/cloud/vmware/vmware_vm_host_drs_rule.py]: Ok, no problem - will fix

File: lib/ansible/modules/cloud/vmware/vmware_vm_host_drs_rule.py (hunk 10 above)
Comment (karstenjakobsen) [reply lib/ansible/modules/cloud/vmware/vmware_vm_host_drs_rule.py]: Ok, no problem - will fix

**Comment by karstenjakobsen** [comment]
Added new review changes - please let me know if anything else is missing :)

**Review by Akasurde — APPROVED** [review]

**Comment by Akasurde** [comment]
@pdellaert @dericcrago @Im0 @tchernomax @ckotte @jeking3 @Tomrrow9 Could you please review this ?

File: lib/ansible/modules/cloud/vmware/vmware_vm_host_drs_rule.py [hunk 11]
```diff
@@ -0,0 +1,470 @@
...
+        ['state', 'present', ['vm_group_name'], ['host_group_name']],
+    ]
+
+    module = AnsibleModule(argument_spec=argument_spec,
+                           required_if=required_if,
+                           supports_check_mode=True)
+
+    try:
+        # Create instance of VmwareDrsGroupManager
+        vm_host_drs = VmwareVmHostRuleDrs(module=module,
```
Comment (pdellaert) [inline lib/ansible/modules/cloud/vmware/vmware_vm_host_drs_rule.py]: Wouldn't it be easier to only have the module parameter here, and not the rest. And in the `init` method, you use the `module.params.get(...`?

Just simplifying the design for maintainability (adding a new argument, would require changes in two places to be consistent)

**Review by pdellaert — APPROVED** [review]
Thanks for this contribution! 

I have one suggestion as an improvement, but don't want to block the PR because of it

File: lib/ansible/modules/cloud/vmware/vmware_vm_host_drs_rule.py (hunk 11 above)
Comment (karstenjakobsen) [reply lib/ansible/modules/cloud/vmware/vmware_vm_host_drs_rule.py]: Yeah, I see what you mean. I think I did it for the purpose of making the class `VmwareVmHostRuleDrs` a tad more agnostic to Ansible. But I can make the changes - no probs :)

Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_vm_host_drs_rule.py]: @karstenjakobsen Let me know once you are done with changes. Thanks.

Comment (karstenjakobsen) [reply lib/ansible/modules/cloud/vmware/vmware_vm_host_drs_rule.py]: Done :)

**Review by pdellaert — APPROVED** [review]
LGTM!

**Pull request merged**